apps/emqx/src/emqx_router_helper.erl

```diff
@@ -188,6 +230,228 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%--------------------------------------------------------------------
 
+handle_membership_event({node, down, Node}, State) ->
+    _ = record_node_down(Node),
+    State;
+handle_membership_event({node, leaving, Node}, State) ->
+    _ = schedule_purge_left(Node),
+    _ = record_node_left(Node),
+    State;
+handle_membership_event({node, up, Node}, State) ->
+    _ = record_node_alive(Node),
+    State;
+handle_membership_event({mnesia, up, Node}, State = #{last_membership := Membership}) ->
+    State#{last_membership := lists:usort([Node | Membership])};
+handle_membership_event(_Event, State) ->
+    State.
+
+record_node_down(Node) ->
```

Comment:
write a warning message ?

Reply:
There's at least https://github.com/emqx/mria/blob/603bea2481917041e502952e834ccc53ddc876f0/src/mria_membership.erl#L340, though I wouldn't be surprised if there are more.

Why warning though? Nodes coming up and down are regular lifecycle events. Nevertheless, if we want to increase visibility, it's probably better to do it on `mria` side?

Reply:
Sorry, maybe not here, but at the place where one node decides to clean-up a peer node.
This is for better traceability in case some routes missing, we can look for this log to see when the routes were cleared by which node.

Reply:
There's https://github.com/emqx/emqx/blob/4c1135eb4bd69dad849a21ce45412dc715ae20e2/apps/emqx/src/emqx_router_helper.erl#L362. There's no event _before_ this is done, might add it as well I guess.

Still unsure about warning level, because the current expectation is that a node is not supposed to completely clean its own routes during orderly shutdown. I was thinking about changing it to "node should clean its routes during shutdown", and then this event indeed can be turned into a warning (because it would then imply node crash / permanent connectivity loss), and thus become more visible. What do you think?

Reply:
Although not documented, a proper shutdown is to evict the node first
in this sense, any abrupt shutdown is considered non-graceful and worth some warning.
Maybe write a `warning` if there is ANY route to this node, and `debug` if ignore

Reply:
>  Maybe write a `warning` if there is ANY route to this node, and `debug` if ignore

It is like that right now.

> Although not documented, a proper shutdown is to evict the node first

With the current expectation it would still produce a warning (because evicted / "left" node don't clean own routes during shutdown).

But in general I can imagine the one who operates the cluster might just want to `systemctl restart emqx` for whatever reason, do you think it still needs to be considered non-graceful shutdown?

Reply:
> It is similar right now, info and debug if node had no routes.

Great!

> Do you think it still needs to be considered non-graceful shutdown?

I don't really want to argue which way of shutdown is or is not graceful.
What I want is some logs to help investigation if something odd happens, and it seems raising it to warning is the simplest way.
The downside is someone may ask why this is a warning, what should I (as a sys-admin) do, we can maybe add more context in the log like: "if this peer node is shutdown by admin, ignore this warning message" ?

Reply:
> raising it to warning is the simplest way.

True. Okay, will raise the event to a warning.

In general though, can't help but argue that it's always the simplest way. So if we always follow the same principle we'll end up with logs sprinkled with warnings that are not actually warnings.

Reply:
Done in 7b5e59f05d2430162261f8057346f908f700e250.